### PR TITLE
PACA-920: Use DSL to generate all RIF parsers (Part 2/2)

### PR DESCRIPTION
**Pull request by brianburton**
**JIRA Ticket:**
[PACA-920](https://jira.cms.gov/browse/PACA-920)

**User Story or Bug Summary:**

As a BFD engineer I want to have a single code generator for both RIF and RDA entities and transformer/parsers.

---

### What Does This PR Do?

The PR is part two of a two sprint effort.  Refer to the [PR for part 1](https://github.com/CMSgov/beneficiary-fhir-data/pull/1546) to see the code and config changes related to hibernate entity generation.

- Removes the RIF code generation annotation processor.
- Removes RIF spreadsheet `rif-layout-and-fhir-mapping.xlsx`
- Adds CsvWriter generation goal to DSL plugin.
- Updates transformations in all RIF mappings to cover all columns.
- Changes output directory property name for all goals to allow a single plugin config in pom to handle all goals.
- Adds generation of static `parseRif()` method for RIF based entities to ease migration from annotation based code to DSL based code.
- Ensures `ValidationUtil` never returns duplicate error messages.

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:

* Verify all PR security questions and checklists have been completed and addressed.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [x] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [x] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [x] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [x] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* N/A

### Submitter Checklist

I have gone through and verified that...:

* [x] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [x] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [x] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [x] The data dictionary has been updated with any field mapping changes, if any were made.
* [x] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [x] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [x] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [x] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

File: apps/bfd-model/bfd-model-dsl-codegen/bfd-model-dsl-codegen-plugin/src/test/resources/GenerateCsvWritersFromDslMojoIT/expected/BeneficiaryCsvWriter.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+public final class BeneficiaryCsvWriter {
+  public static Map<String, Object[][]> toCsvRecordsByTable(Beneficiary entity) {
+    // Verify the input.
+    Objects.requireNonNull(entity);
+
+    Map<String, Object[][]> csvRecordsByTable = new HashMap<>(2);
+
+    // Convert the header fields.
+    Object[][] headerRecords = new Object[2][];
+    headerRecords[0] = new Object[]{ "BENE_ID", "STATE_CODE", "BENE_SEX_IDENT_CD", "BENE_ENTLMT_RSN_ORIG", "BENE_MDCR_STATUS_CD", "EFCTV_END_DT", "BENE_LINK_KEY", "HICN_UNHASHED", "MBI_HASH", "LAST_UPDATED" };
```
Comment (mackec1) [inline apps/bfd-model/bfd-model-dsl-codegen/bfd-model-dsl-codegen-plugin/src/test/resources/GenerateCsvWritersFromDslMojoIT/expected/BeneficiaryCsvWriter.java:18]: I didn't touch these (as yet); will do shortly.

**Review by mackec1 — APPROVED** [review]
I'll dig into the ITs this weekend; let me know if we should wait til post-merge or on this branch? This stuff looks awesome...can't wait to get rid of .xslx...burned burling and me yesterday!

File: apps/bfd-model/bfd-model-dsl-codegen/bfd-model-dsl-codegen-plugin/src/main/java/gov/cms/model/dsl/codegen/plugin/GenerateCsvWritersFromDslMojo.java [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+    for (ColumnBean column : mapping.getTable().getColumns()) {
+      if (column.hasSequence()) {
+        // WORK AROUND: For some reason the annotation processor code does not print a header for
+        // the sequence id but does include its value so we simulate that here.
+        continue;
+      }
+      if (column.isDbOnly()) {
+        // skip columns that aren't actually properties of the entity
+        continue;
+      }
```
Comment (daryll) [inline apps/bfd-model/bfd-model-dsl-codegen/bfd-model-dsl-codegen-plugin/src/main/java/gov/cms/model/dsl/codegen/plugin/GenerateCsvWritersFromDslMojo.java:294]: cool.
we dont combine these to be very explicit...correct ?

**Review by KyRobbins — COMMENTED** [review]
Looks good, pending the test updates

File: apps/bfd-model/bfd-model-rif-samples/src/main/resources/rif-synthea/beneficiary_2011.csv [hunk 3]
```diff
@@ -1,4 +1,4 @@
...
...BP_NOV_ID|PTD_PBP_DEC_ID|PTD_SGMT_JAN_ID|PTD_SGMT_FEB_ID|PTD_SGMT_MAR_ID|PTD_SGMT_APR_ID|PTD_SGMT_MAY_ID|PTD_SGMT_JUN_ID|PTD_SGMT_JUL_ID|PTD_SGMT_AUG_ID|PTD_SGMT_SEPT_ID|PTD_SGMT_OCT_ID|PTD_SGMT_NOV_ID|PTD_SGMT_DEC_ID|RDS_JAN_IND|RDS_FEB_IND|RDS_MAR_IND|RDS_APR_IND|RDS_MAY_IND|RDS_JUN_IND|RDS_JUL_IND|RDS_AUG_IND|RDS_SEPT_IND|RDS_OCT_IND|RDS_NOV_IND|RDS_DEC_IND|META_DUAL_ELGBL_STUS_JAN_CD|META_DUAL_ELGBL_STUS_FEB_CD|META_DUAL_ELGBL_STUS_MAR_CD|META_DUAL_ELGBL_STUS_APR_CD|META_DUAL_ELGBL_STUS_MAY_CD|META_DUAL_ELGBL_STUS_JUN_CD|META_DUAL_ELGBL_STUS_JUL_CD|META_DUAL_ELGBL_STUS_AUG_CD|META_DUAL_ELGBL_STUS_SEPT_CD|META_DUAL_ELGBL_STUS_OCT_CD|META_DUAL_ELGBL_STUS_NOV_CD|META_DUAL_ELGBL_STUS_DEC_CD|CST_SHR_GRP_JAN_CD|CST_SHR_GRP_FEB_CD|CST_SHR_GRP_MAR_CD|CST_SHR_GRP_APR_CD|CST_SHR_GRP_MAY_CD|CST_SHR_GRP_JUN_CD|CST_SHR_GRP_JUL_CD|CST_SHR_GRP_AUG_CD|CST_SHR_GRP_SEPT_CD|CST_SHR_GRP_OCT_CD|CST_SHR_GRP_NOV_CD|CST_SHR_GRP_DEC_CD|DRVD_LINE_1_ADR|DRVD_LINE_2_ADR|DRVD_LINE_3_ADR|DRVD_LINE_4_ADR|DRVD_LINE_5_ADR|DRVD_LINE_6_ADR|CITY_NAME|STATE_CD|STATE_CNTY_ZIP_CD|EFCTV_BGN_DT|EFCTV_END_DT|BENE_LINK_KEY|PTA_CVRG_STRT_DT|PTA_CVRG_END_DT|PTB_CVRG_STRT_DT|PTB_CVRG_END_DT|PTD_CVRG_STRT_DT|PTD_CVRG_END_DT
```
Comment (KyRobbins) [inline apps/bfd-model/bfd-model-rif-samples/src/main/resources/rif-synthea/beneficiary_2011.csv:1]: Is there a test in the master branch that will get updated to check for these new fields so we can compare its results with this branch's changes?

Comment (brianburton) [reply apps/bfd-model/bfd-model-rif-samples/src/main/resources/rif-synthea/beneficiary_2011.csv]: The new fields have been added to master and this branch has been updated with those changes so all tests are once again enabled and pass.

**Review by KyRobbins — APPROVED** [review]

File: apps/bfd-model/bfd-model-dsl-codegen/bfd-model-dsl-codegen-plugin/src/test/java/gov/cms/model/dsl/codegen/plugin/transformer/ShortStringFieldTransformerTest.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+            .transformation(transformation)
+            .build();
+
+    ShortStringFieldTransformer generator = new ShortStringFieldTransformer();
+    CodeBlock block =
+        generator.generateCodeBlock(
+            mapping, column, transformation, GrpcGetter.Instance, StandardSetter.Instance);
+    assertEquals(
+        "transformer.copyShortString(namePrefix + gov.cms.test.Entity.Fields.idrDtlCnt, true, from.getIdrDtlCnt(), to::setIdrDtlCnt);\n",
+        block.toString());
```
Comment (lsmitchell) [inline apps/bfd-model/bfd-model-dsl-codegen/bfd-model-dsl-codegen-plugin/src/test/java/gov/cms/model/dsl/codegen/plugin/transformer/ShortStringFieldTransformerTest.java:37]: I'd wonder if this isn't a bit brittle; perhaps we'd be better just checking that "copyShortString" exists within the generated code block?

File: apps/bfd-model/bfd-model-dsl-codegen/bfd-model-dsl-codegen-plugin/src/test/resources/GenerateCsvWritersFromDslMojoIT/expected/BeneficiaryCsvWriter.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
+package gov.cms.test;
+
+import java.lang.Object;
+import java.lang.String;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+public final class BeneficiaryCsvWriter {
```
Comment (lsmitchell) [inline apps/bfd-model/bfd-model-dsl-codegen/bfd-model-dsl-codegen-plugin/src/test/resources/GenerateCsvWritersFromDslMojoIT/expected/BeneficiaryCsvWriter.java:9]: Missing various docs in this class (Checkstyle should be catching this...)

File: apps/bfd-model/bfd-model-dsl-codegen/bfd-model-dsl-codegen-plugin/src/test/resources/GenerateCsvWritersFromDslMojoIT/expected/DMEClaimCsvWriter.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
+package gov.cms.test;
+
+import java.lang.Object;
+import java.lang.String;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+public final class DMEClaimCsvWriter {
```
Comment (lsmitchell) [inline apps/bfd-model/bfd-model-dsl-codegen/bfd-model-dsl-codegen-plugin/src/test/resources/GenerateCsvWritersFromDslMojoIT/expected/DMEClaimCsvWriter.java:9]: Docs missing here too; looked into this and checkstyle is missing this cause this java file is in test/resources. IMO source files shouldnt be in resources; if this is a utility class to create expected responses, it should still be under test/java, unless there is some compiler reason this is required to be here.

**Review by lsmitchell — COMMENTED** [review]
Just a couple things! Excited to get this going.

File: apps/bfd-model/bfd-model-dsl-codegen/bfd-model-dsl-codegen-plugin/src/test/java/gov/cms/model/dsl/codegen/plugin/transformer/ShortStringFieldTransformerTest.java (hunk 4 above)
Comment (brianburton) [reply apps/bfd-model/bfd-model-dsl-codegen/bfd-model-dsl-codegen-plugin/src/test/java/gov/cms/model/dsl/codegen/plugin/transformer/ShortStringFieldTransformerTest.java:37]: Just verifying the string is generated correctly.  Same as `assertEquals(1)` would be for an integer field.

File: apps/bfd-model/bfd-model-dsl-codegen/bfd-model-dsl-codegen-plugin/src/test/resources/GenerateCsvWritersFromDslMojoIT/expected/BeneficiaryCsvWriter.java (hunk 5 above)
Comment (brianburton) [reply apps/bfd-model/bfd-model-dsl-codegen/bfd-model-dsl-codegen-plugin/src/test/resources/GenerateCsvWritersFromDslMojoIT/expected/BeneficiaryCsvWriter.java:9]: This isn't a compiled class.  It's an expected output from the code generator.  If checkstyle was configured to test it that would be a bug.

File: apps/bfd-model/bfd-model-dsl-codegen/bfd-model-dsl-codegen-plugin/src/test/resources/GenerateCsvWritersFromDslMojoIT/expected/DMEClaimCsvWriter.java (hunk 6 above)
Comment (brianburton) [reply apps/bfd-model/bfd-model-dsl-codegen/bfd-model-dsl-codegen-plugin/src/test/resources/GenerateCsvWritersFromDslMojoIT/expected/DMEClaimCsvWriter.java:9]: Nothing missing here.  This is an expected output file for the test.

**Review by mjburling — APPROVED** [review]

**Review by KyRobbins — APPROVED** [review]

**Pull request merged**